Add Django settings flags to roll out the extracted XBlocks

farhan opened the pull request:
Add Waffle flags to roll out the extracted XBlocks 
Flags will use to toggle between the old and new block quickly
without putting course content or user state at risk.

Ticket: https://github.com/openedx/edx-platform/issues/35308

### Django Settings Flags
Following built in blocks wafffle flags are in action in this PR and has been implemented:

- [x] word_cloud
- [x] annotatable
- [x] poll_question
- [x] lti
- [x] html
- [x] discussion
- [x] problem
- [x] video

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281277244.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281277244-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281277244-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281277244-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281415951.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281415951-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281415951-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281415951-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281700486.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281700486-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281700486-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8281700486-tutor-requirements.txt)

farhan committed:
chore: Add xblocks-contrib in budle.in

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-35549-139931.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-35549-139931.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8294828010-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8294828010-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8294828010-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-35549-139931.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-35549-139931.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8295282265-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8295282265-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8295282265-tutor-requirements.txt)

kdmccormick commented on xmodule/toggles.py:
```diff
@@ -0,0 +1,23 @@
+USE_EXTRACTED_WORD_CLOUD_BLOCK = WaffleFlag('xmodule.use_extracted_block.word_cloud', __name__)
+# TODO: Add flag docs once above approved
+USE_EXTRACTED_ANNOTATABLE_BLOCK = WaffleFlag('xmodule.use_extracted_block.annotatable', __name__)
+USE_EXTRACTED_POLL_BLOCK = WaffleFlag('xmodule.use_extracted_block.poll', __name__)
```
```suggestion
USE_EXTRACTED_POLL_QUESTION_BLOCK = WaffleFlag('xmodule.use_extracted_block.poll_question', __name__)
```

This suggestion would make it so the flag matches the `poll_question` entrypoint name in setup.py, and help avoid confusing it with [OpenCraft's `poll` block](https://github.com/open-craft/xblock-poll).

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-35549-139931.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-35549-139931.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8304614839-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8304614839-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8304614839-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-35549-139931.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-35549-139931.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8304872328-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8304872328-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-35549-139931-51564656-8304872328-tutor-requirements.txt)

kdmccormick commented:
Looking good so far. Let's keep this marked as a Draft until the changeset is ready for review.

farhan committed:
feat: Upgrade Python dependency xblocks-contrib

chore: Update xblocks-contrib pypi package

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/farhan/waffle-flag-for-extracted-xblock`

farhan commented:
@kdmccormick 
Please review the initial steps of this PR and guide us either we are in right direction, specially how we are fixing test cases.

cc: @irtazaakram @ttqureshi

kdmccormick commented:
Thanks for the ping @farhan ... reviewing your unit test fixes, it's become clear that my suggested approach is flawed 😓 

It seems that we cannot use Waffle flags at the level of the module, because Django is not yet ready to access the database at module-load time. You have figured out a clever workaround for unit tests, but this would not solve the issue at runtime.

I am experimenting with some alternatives right now. I'll let you know what I find.

kdmccormick commented:
@farhan I think that we need to use regular boolean Django settings (e.g.,`settings.USE_EXTRACTED_WORD_CLOUD_BLOCK`) rather than Waffle flags. This will be somewhat less flexible for site operators, but reading the flags won't require database access, so it should resolve these startup errors that you are seeing with the Waffle flags. Can you try converting these to boolean Django settings and see if the original (unchanged)  unit tests pass?

farhan committed:
chore: add analytics

farhan commented:
@kdmccormick @openedx/axim-aximprovements @ttqureshi  PR is available for code review
Things to mention:
1. Please review the settings files, either I have added the flags in right module
2. Tests are failing https://github.com/openedx/edx-platform/pull/35549/commits/9b75f1e65b7e990a7a2df20f71df8ba71572fcc4 when turn the extracted block flags, which is fine but should be tested when we will publish the extracted block functionality in the `xblocks-contrib` pypi package.
     a. We should add it in the acceptance criteria of the extracted xblock story.

ttqureshi reviewed (commented):
Looks good to me!

kdmccormick commented on lms/envs/common.py:
```diff
@@ -72,6 +72,16 @@
+    USE_EXTRACTED_DISCUSSION_BLOCK,
+    USE_EXTRACTED_PROBLEM_BLOCK,
+    USE_EXTRACTED_VIDEO_BLOCK,
+)
```
Hm, I never seen this pattern before. Is there a need to define these flags in xmodule/toggles.py, or could we just define them in lms/envs/common.py like the rest of the settings?

farhan replied:
resolved

kdmccormick commented on xmodule/toggles.py:
```diff
@@ -0,0 +1,87 @@
+# .. toggle_name: USE_EXTRACTED_WORD_CLOUD_BLOCK
+# .. toggle_default: False
+# .. toggle_implementation: DjangoSetting
+# .. toggle_description: Enables the use of the extracted Word Cloud XBlock, which has been shifted to the 'openedx/xblocks-contrib' repo.
```
Great descriptions on all of these, thank you!

kdmccormick commented on xmodule/tests/test_vertical.py:
```diff
@@ -89,7 +90,7 @@ class TestVerticalBlockChildRenderStep(PipelineStep):
     def run_filter(self, block, context):  # lint-amnesty, pylint: disable=arguments-differ
         """Pipeline step that changes child content"""
-        if type(block).__name__ == "HtmlBlockWithMixins":
+        if type(block).__name__ == f"{HtmlBlock.__name__}WithMixins":
```
Nice fix.

kdmccormick reviewed (changes requested):
Thanks for converting the flags to toggles. Just a couple more comments from me.

farhan commented:
@kdmccormick @feanil 
PR is available for code review again
CMS requires to add the settings flags into its `cms/envs/common.py` file, so I have imported there as well
Tests were failing https://github.com/openedx/edx-platform/pull/35549/commits/2b57c296a0e53f6ac0aa81efc4ef913e72e3ca33

Please squash & merge in time slot, if all seems ok

feanil reviewed (commented):
It looks like you had to correct various tests because the `__name__` of the class changed. Is the name of the class also going to show up in exports or other serializations? If so, we should not update the name here or update the `__name__` back to preserve the name in serializations.

farhan commented:
> It looks like you had to correct various tests because the `__name__` of the class changed. Is the name of the class also going to show up in exports or other serializations? If so, we should not update the name here or update the `__name__` back to preserve the name in serializations.

Thanks for sharing your extensive thoughts @feanil 

On my recent testing experience with some Problem XBlocks, their UI is disturbed, it's fine on master

![screencapture-apps-local-openedx-io-2000-learning-course-course-v1-edx-2-2-block-v1-edx-2-2-type-sequential-block-182eadafb52f4a99b203cf8429730fe7-block-v1-edx-2-2-type-vertical-block-fa5e73a74e5d44d49e6323da495acd1d-2024-11-26-18_45_01](https://github.com/user-attachments/assets/739e3ac4-e1c5-47fa-8201-c42cef8cc34b)

Let's figure out the ripples and have its testing
cc: @irtazaakram @ttqureshi

kdmccormick commented:
> It looks like you had to correct various tests because the __name__ of the class changed. Is the name of the class also going to show up in exports or other serializations? If so, we should not update the name here or update the __name__ back to preserve the name in serializations.

This is a good thing to keep in mind when changing the names of any classes which are persisted to the DB. In this particular case, though, I think we are OK. As far as I know, there are no features that depend on the literal class name of the block. The names that are used in de/serialization are the tag name(s) (`problem`, `video`, etc.) that each block is mapped to in setup.py.

Looking through the test cases which were updated, it seems that they were checking the class name rather than the class itself only because the class itself (`BlahBlockWithMixins`) is generated on-the-fly by mixologist.

kdmccormick commented:
@farhan Taking a look at your screenshot, it seems that either some CSS or JS (or both) is missing. And if you are able to still submit the problem's answer successfully, then that would imply that the JS is present, and that only the CSS is missing.

If you haven't already, try re-building static assets (`npm run build-dev` within `tutor dev run lms bash`). If that doesn't work, then I imagine that we have broke something which maps the builtin block classes to their corresponding static assets.

kdmccormick commented:
One fix you could try, related to @feanil 's comment earlier, would be setting `__name__` on the built-in class after its definition:
```python
_BuiltInProblemBlock.__name__ = "ProblemBlock"
```
This essentially tells Python "From the perspective of the module, the class goes by one name (_BuiltInProblemBlock), but I want it to appear to other objects as if it has a different name (ProblemBlock)".

farhan commented:
@kdmccormick Thanks for sharing thoughts

1. Related to exporting of course, xblocks are exporting fine as they are using tags rather than xblock class names
2. I have checked again even rebuilding assets via `npm run build` is not fixing UI issue (Problems, Video)
3. Running `./manage.py lms collectstatic ` is not fixing the ui either
4. CSS files are generated and present but they are not linking correctly or reading css from the required files
5. Problem is submitting and checked status exist even on reloading page
6. When I run the same assets on master branch, UI/styling fixes.

Next:
- So my as per my analysis, we needs to look into why css files are not linking correctly after changing of class names.
- Check the effectiveness of following solution:
```_BuiltInProblemBlock.__name__ = "ProblemBlock"```

kdmccormick commented:
Sounds good @farhan . You may already know this, but just in case: if you are running on `tutor dev`, then you will want to use `npm run build-dev` rather than `npm run build`. If you are running `tutor local`, then `npm run build` is appropriate.

farhan commented:
> Sounds good @farhan . You may already know this, but just in case: if you are running on `tutor dev`, then you will want to use `npm run build-dev` rather than `npm run build`. If you are running `tutor local`, then `npm run build` is appropriate.

I had tried `npm run build-dev` as well.

farhan committed:
chore: Adds django settings flag impl

farhan committed:
chore: fix test cases

farhan committed:
chore: change class name defining strategy for fix test cases

farhan committed:
Revert "chore: fix test cases"

This reverts commit 948922c0e43b3ff7a31f1efbde6abf50c42ca0c9.

farhan commented:
PR is available for code review.
Problem and Video Blocks are rendering normal after following change.
https://github.com/openedx/edx-platform/pull/35549/files#diff-2adcf1a5ddc155e832aeab0472da97156e8487a683da9f786198332367eacc20R2519

I have reverted the test case fixing commit.

kdmccormick commented on xmodule/html_block.py:
```diff
@@ -498,3 +498,4 @@ def safe_parse_date(date):
     _ExtractedHtmlBlock if settings.USE_EXTRACTED_HTML_BLOCK
     else _BuiltInHtmlBlock
 )
+HtmlBlock.__name__ = "HtmlBlock"
```
We need to ensure that the HTML subclasses (CourseInfoBlock, AboutBlock, and StaticTabBlock) are based on the extracted HTML block code rather than the built-in code, as they are now. This may require exposing a new HTMLMixin class from xblocks-contrib.

This can be done in a follow-up.

kdmccormick reviewed (approved):
Good work. I have not tested it myself but I see that you have.
